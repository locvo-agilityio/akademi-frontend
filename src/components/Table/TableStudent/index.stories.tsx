import { StoryObj, Meta } from '@storybook/react';
import { Avatar, Box, Flex, Td, Text } from '@chakra-ui/react';

// Components
import TableStudent from '.';

const meta: Meta<typeof TableStudent> = {
  title: 'Components/Tables/TableStudent',
  tags: ['autodocs'],
  component: TableStudent,
  render: (props) => (
    <Flex
      w="full"
      h="100vh"
      bgColor="gray.150"
      alignItems="center"
      justifyContent="center"
    >
      <Box p={8} borderRadius="lg" w="1140px">
        <TableStudent {...props} />
      </Box>
    </Flex>
  ),
};

export default meta;
type Story = StoryObj<typeof TableStudent>;

export const Default: Story = {
  args: {
    columns: [
      {
        title: 'Name',
        key: 'name',
        renderBody: ({ firstName, lastName, avatar }) => (
          <Td display="flex" p={0} border="none" alignItems="center" gap={4}>
            <Avatar
              size="md"
              name={`${firstName} ${lastName}`}
              src={avatar as string}
            />
            <Text fontWeight="bold" color="darkBlue">
              {firstName} {lastName}
            </Text>
          </Td>
        ),
      },
      {
        title: 'ID',
        key: 'id',
      },
      {
        title: 'Date',
        key: 'date',
      },
      {
        title: 'Parent Name',
        key: 'parentName',
      },
      {
        title: 'City',
        key: 'city',
      },
      {
        title: 'Contact',
        key: 'contact',
      },
      {
        title: 'Grade',
        key: 'grade',
      },
      {
        title: 'Action',
        key: 'action',
      },
    ],
    dataSource: [
      {
        id: '1',
        avatar: 'https://bit.ly/dan-abramov',
        firstName: 'John',
        lastName: 'Doe',
        date: '2022-01-01',
        parentName: 'John Doe',
        city: 'New York',
        contact: '1234567890',
        grade: 'A',
      },
    ],
  },
};
