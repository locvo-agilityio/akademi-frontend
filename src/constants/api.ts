import { TQueryKey } from '@/types';

export const API_PATHS = {
  TEACHERS: '/teachers',
  STUDENTS: '/students',
};

export const studentsQueryKeys = {
  all: [{ scope: 'students' }] as const,
  lists: () => [{ ...studentsQueryKeys.all[0], entity: 'list' }] as const,
  list: ({ page, filter }: TQueryKey) =>
    [
      {
        ...studentsQueryKeys.lists()[0],
        page,
        ...(filter && { filter }),
      },
    ] as const,
  details: () => [{ ...studentsQueryKeys.all[0], entity: 'detail' }] as const,
  detail: (id?: string) => [{ ...studentsQueryKeys.details()[0], id }] as const,
};

export const teachersQueryKeys = {
  all: [{ scope: 'teachers' }] as const,
  lists: () => [{ ...teachersQueryKeys.all[0], entity: 'list' }] as const,
  list: ({ page, filter }: TQueryKey) =>
    [
      {
        ...teachersQueryKeys.lists()[0],
        page,
        ...(filter && { filter }),
      },
    ] as const,
  details: () => [{ ...teachersQueryKeys.all[0], entity: 'detail' }] as const,
  detail: (id?: string) => [{ ...teachersQueryKeys.details()[0], id }] as const,
};
