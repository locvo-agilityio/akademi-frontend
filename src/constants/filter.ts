export const FILTER_OPTIONS = [
  {
    value: 'desc',
    label: 'Descending',
  },
  {
    value: 'asc',
    label: 'Ascending',
  },
];
