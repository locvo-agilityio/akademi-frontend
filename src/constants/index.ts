export * from './sidebar';
export * from './message';
export * from './router';
export * from './filter';
export * from './httpsRequest';
export * from './api';
export * from './query';
export * from './pagination';
export * from './teacher';
export * from './file';
export * from './form';
export * from './regex';
export * from './status';
export * from './student';
