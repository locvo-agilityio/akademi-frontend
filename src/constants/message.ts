export const ERROR_MESSAGES = {
  EMPTY_DATA: 'No data available',
  MAX_SIZE_IMAGE: 'Image size must be less than 3MB',
  INVALID_FILE_TYPE: 'File must be a PNG, JPEG, or JPG',
  FIELD_REQUIRED: (fieldName: string) => `${fieldName} is required`,
  INVALID_PHONE: 'Invalid US phone number format',
  INVALID_EMAIL: 'Please enter a valid email address',
  INVALID_MIN_LENGTH: 'Please enter at least 2 characters',
  INVALID_UPPERCASE_FIRST_LETTER: 'The first letter must be uppercase',
  ADD_TEACHER_FAILED: 'Add teacher failed',
  ADD_STUDENT_FAILED: 'Add student failed',
  UPLOAD_IMAGE_FAILED: 'Upload image failed',
  DELETE_STUDENT_FAILED: 'Delete student failed',
  DELETE_TEACHER_FAILED: 'Delete teacher failed',
  EDIT_TEACHER_FAILED: 'Edit teacher failed',
  EDIT_STUDENT_FAILED: 'Edit student failed',
};

export const SUCCESS_MESSAGES = {
  ADD_TEACHER_SUCCESS: 'Add teacher successfully',
  ADD_STUDENT_SUCCESS: 'Add student successfully',
  DELETE_STUDENT_SUCCESS: 'Delete student successfully',
  DELETE_TEACHER_SUCCESS: 'Delete teacher successfully',
  EDIT_TEACHER_SUCCESS: 'Edit teacher successfully',
  EDIT_STUDENT_SUCCESS: 'Edit student successfully',
  UPLOAD_IMAGE_SUCCESS: 'Upload image successfully',
};
