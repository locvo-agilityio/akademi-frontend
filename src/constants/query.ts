export const QUERY = {
  STUDENTS_UNPAID: (page: number) =>
    `filters[payment]=unpaid&pagination[page]=${page}&pagination[pageSize]=5&sort=createdAt:desc`,
  STUDENTS: (page: number) => `pagination[page]=${page}&pagination[pageSize]=5`,
  TEACHERS: (page: number) =>
    `pagination[page]=${page}&pagination[pageSize]=10`,
};
