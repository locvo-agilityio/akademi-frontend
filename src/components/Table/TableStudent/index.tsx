import { memo, useMemo } from 'react';
import {
  TableContainer,
  Table as TableChakra,
  Thead,
  Tbody,
  TableProps,
  Tr,
  Td,
} from '@chakra-ui/react';
import isEqual from 'react-fast-compare';

// Types
import { TDataSource, THeaderTable } from '@/types';

// Components
import TableRow from '@/components/Table/TableRow';
import TableHeader from '@/components/Table/TableHeader';
import Fallback from '@/components/Fallback';

// Constants
import { ERROR_MESSAGES } from '@/constants';

import { processTableData } from '@/utils';

type TTableStudentProps = TableProps & {
  isLoading?: boolean;
  columns?: THeaderTable[];
  dataSource?: TDataSource[];
};

const TableStudent = ({
  isLoading,
  columns = [],
  dataSource = [],
  ...props
}: TTableStudentProps) => {
  const processedData = useMemo(
    () => processTableData(columns, dataSource),
    [columns, dataSource],
  );

  return (
    <TableContainer
      w="full"
      h="full"
      overflowY="scroll"
      css={{
        '&::-webkit-scrollbar': {
          width: 2,
        },
        '&::-webkit-scrollbar-track': {
          width: 2,
        },
        '&::-webkit-scrollbar-thumb': {
          background: 'gray',
          borderRadius: '24px',
        },
      }}
    >
      <TableChakra {...props}>
        <Thead>
          <TableHeader columns={columns} />
        </Thead>

        <Tbody>
          {processedData.length === 0 ? (
            <Tr>
              <Td
                colSpan={columns?.length}
                textAlign="center"
                fontWeight="semibold"
                border="none"
                fontSize="lg"
              >
                {isLoading ? <Fallback /> : ERROR_MESSAGES.EMPTY_DATA}
              </Td>
            </Tr>
          ) : (
            processedData.map((data) => (
              <TableRow key={data.id} cells={data.cells} />
            ))
          )}
        </Tbody>
      </TableChakra>
    </TableContainer>
  );
};

export default memo(TableStudent, isEqual);
