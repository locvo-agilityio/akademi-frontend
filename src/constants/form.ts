import { ACCEPTED_FILE_TYPES, MAX_UPLOAD_SIZE } from './file';
import { ERROR_MESSAGES } from './message';
import { REGEX } from './regex';

// Types
import { IParent, IStudent, ITeacher, TField } from '@/types';

export const VALIDATION_RULES = {
  IMAGE: {
    validate: (file: File) => {
      if (file.size > MAX_UPLOAD_SIZE) {
        return ERROR_MESSAGES.MAX_SIZE_IMAGE;
      }

      if (!ACCEPTED_FILE_TYPES.includes(file.type)) {
        return ERROR_MESSAGES.INVALID_FILE_TYPE;
      }
    },
  },
  FIRST_NAME: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('First name'),
    pattern: {
      value: REGEX.UPPERCASE,
      message: ERROR_MESSAGES.INVALID_UPPERCASE_FIRST_LETTER,
    },
  },
  LAST_NAME: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Last name'),
    pattern: {
      value: REGEX.UPPERCASE,
      message: ERROR_MESSAGES.INVALID_UPPERCASE_FIRST_LETTER,
    },
  },
  BIRTH_DATE: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Date of birth'),
  },
  CITY: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('City'),
  },
  EMAIL: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Email'),
    pattern: {
      value: REGEX.EMAIL,
      message: ERROR_MESSAGES.INVALID_EMAIL,
    },
  },
  ADDRESS: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Address'),
  },
  PARENT_NAME: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Parent name'),
    pattern: {
      value: REGEX.UPPERCASE,
      message: ERROR_MESSAGES.INVALID_UPPERCASE_FIRST_LETTER,
    },
  },
  PHONE: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Phone'),
    pattern: {
      value: REGEX.PHONE_NUMBER,
      message: ERROR_MESSAGES.INVALID_PHONE,
    },
  },
  DEGREE: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Degree'),
    pattern: {
      value: REGEX.UPPERCASE,
      message: ERROR_MESSAGES.INVALID_UPPERCASE_FIRST_LETTER,
    },
  },
  UNIVERSITY: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('University'),
    pattern: {
      value: REGEX.UPPERCASE,
      message: ERROR_MESSAGES.INVALID_UPPERCASE_FIRST_LETTER,
    },
  },
  SUBJECT: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Subject'),
    pattern: {
      value: REGEX.UPPERCASE,
      message: ERROR_MESSAGES.INVALID_UPPERCASE_FIRST_LETTER,
    },
  },
  EXPERTISE: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Expertise'),
    pattern: {
      value: REGEX.UPPERCASE,
      message: ERROR_MESSAGES.INVALID_UPPERCASE_FIRST_LETTER,
    },
  },
  DESCRIPTION: {
    pattern: {
      value: REGEX.UPPERCASE,
      message: ERROR_MESSAGES.INVALID_UPPERCASE_FIRST_LETTER,
    },
  },
  DATE: {
    required: ERROR_MESSAGES.FIELD_REQUIRED('Date'),
  },
};

export const ADDITIONAL_TEACHER_INFO = [
  {
    name: 'lastName',
    label: 'Last Name *',
    placeholder: 'Please enter your last name',
    type: 'text',
    rules: VALIDATION_RULES.LAST_NAME,
    ariaLabel: 'lastName',
  },
  {
    name: 'phone',
    label: 'Phone *',
    placeholder: '(xxx) xxx-xxxx',
    type: 'text',
    rules: VALIDATION_RULES.PHONE,
    ariaLabel: 'phone',
  },
] as Array<TField<Pick<ITeacher, 'lastName' | 'phone'>>>;

export const EDUCATION_TEACHER_INFO = [
  {
    name: 'degree',
    label: 'Degree *',
    placeholder: 'Please enter your degree',
    type: 'text',
    rules: VALIDATION_RULES.DEGREE,
    ariaLabel: 'degree',
  },
  {
    name: 'cityEducation',
    label: 'City *',
    placeholder: 'Please enter your city',
    type: 'text',
    rules: VALIDATION_RULES.CITY,
    ariaLabel: 'cityEducation',
  },
] as Array<TField<Pick<ITeacher, 'degree' | 'cityEducation'>>>;

export const BASIC_PARENT_STUDENT_INFO = [
  {
    name: 'parent.firstName',
    label: 'First Name *',
    placeholder: 'Please enter your first name',
    type: 'text',
    rules: VALIDATION_RULES.FIRST_NAME,
    ariaLabel: 'firstName',
  },
  {
    name: 'parent.email',
    label: 'Email *',
    placeholder: 'Please enter your email',
    type: 'email',
    rules: VALIDATION_RULES.EMAIL,
    ariaLabel: 'email',
  },
] as unknown as Array<TField<Pick<IParent, 'firstName' | 'email'>>>;

export const ADDITIONAL_PARENT_STUDENT_INFO = [
  {
    name: 'parent.lastName',
    label: 'Last Name *',
    placeholder: 'Please enter your last name',
    type: 'text',
    rules: VALIDATION_RULES.LAST_NAME,
    ariaLabel: 'lastName',
  },
  {
    name: 'parent.phone',
    label: 'Phone *',
    placeholder: '(xxx) xxx-xxxx',
    type: 'text',
    rules: VALIDATION_RULES.PHONE,
    ariaLabel: 'parentPhone',
  },
] as unknown as Array<TField<Pick<IParent, 'lastName' | 'phone'>>>;

export const BASIC_STUDENT_INFO = [
  {
    name: 'lastName',
    label: 'Last Name *',
    placeholder: 'Please enter your last name',
    type: 'text',
    rules: VALIDATION_RULES.LAST_NAME,
    ariaLabel: 'lastName',
  },
  {
    name: 'parentName',
    label: 'Parent Name *',
    placeholder: 'Please enter your parent name',
    type: 'text',
    rules: VALIDATION_RULES.PARENT_NAME,
    ariaLabel: 'parentName',
  },
  {
    name: 'phone',
    label: 'Phone *',
    placeholder: '(xxx) xxx-xxxx',
    type: 'text',
    rules: VALIDATION_RULES.PHONE,
    ariaLabel: 'phone',
  },
] as Array<TField<Pick<IStudent, 'lastName' | 'parentName' | 'phone'>>>;

export const BASIC_TEACHER_INFO = [
  {
    name: 'firstName',
    label: 'First Name *',
    placeholder: 'Please enter your first name',
    type: 'text',
    rules: VALIDATION_RULES.FIRST_NAME,
    ariaLabel: 'firstName',
  },
  {
    name: 'email',
    label: 'Email *',
    placeholder: 'Please enter your email',
    type: 'email',
    rules: VALIDATION_RULES.EMAIL,
    ariaLabel: 'email',
  },
  {
    name: 'address',
    label: 'Address',
    placeholder: 'Please enter your address',
    type: 'text',
    rules: VALIDATION_RULES.ADDRESS,
    ariaLabel: 'address',
  },
  {
    name: 'birthday',
    label: 'Date of Birth *',
    placeholder: 'Please enter your birthday',
    min: '1970-12-31',
    max: '2000-12-31',
    type: 'date',
    rules: VALIDATION_RULES.BIRTH_DATE,
    ariaLabel: 'birthday',
  },
  {
    name: 'subject',
    label: 'Subject',
    placeholder: 'Please enter your subject',
    type: 'text',
    rules: VALIDATION_RULES.CITY,
    ariaLabel: 'subject',
  },
  {
    name: 'description',
    label: 'Description',
    placeholder: 'Please enter your description',
    type: 'text',
    rules: VALIDATION_RULES.DESCRIPTION,
    ariaLabel: 'description',
  },
] as Array<
  TField<
    Pick<
      ITeacher,
      'firstName' | 'email' | 'address' | 'birthday' | 'subject' | 'description'
    >
  >
>;
