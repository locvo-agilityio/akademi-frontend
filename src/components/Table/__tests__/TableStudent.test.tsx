import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Avatar, Td, Text } from '@chakra-ui/react';

// Components
import TableStudent from '../TableStudent';
import { ERROR_MESSAGES } from '@/constants';

describe('TableStudent Component', () => {
  const renderSetup = () =>
    render(
      <BrowserRouter>
        <TableStudent
          columns={[
            {
              title: 'Name',
              key: 'name',
              renderBody: ({ firstName, lastName, avatar }) => (
                <Td
                  display="flex"
                  p={0}
                  border="none"
                  alignItems="center"
                  gap={4}
                >
                  <Avatar
                    size="md"
                    name={`${firstName} ${lastName}`}
                    src={avatar as string}
                  />
                  <Text fontWeight="bold" color="darkBlue">
                    {firstName} {lastName}
                  </Text>
                </Td>
              ),
            },
            {
              title: 'ID',
              key: 'id',
            },
            {
              title: 'Date',
              key: 'date',
            },
            {
              title: 'Parent Name',
              key: 'parentName',
            },
            {
              title: 'City',
              key: 'city',
            },
            {
              title: 'Contact',
              key: 'contact',
            },
            {
              title: 'Grade',
              key: 'grade',
            },
            {
              title: 'Action',
              key: 'action',
            },
          ]}
          dataSource={[
            {
              id: '1',
              avatar: 'https://bit.ly/dan-abramov',
              firstName: 'John',
              lastName: 'Doe',
              date: '2022-01-01',
              parentName: 'John Doe',
              city: 'New York',
              contact: '1234567890',
              grade: 'A',
            },
          ]}
        />
      </BrowserRouter>,
    );

  it('Should match snapshot', () => {
    const { container } = renderSetup();
    expect(container).toMatchSnapshot();
  });

  it('should render empty data', () => {
    const { getByText } = render(
      <BrowserRouter>
        <TableStudent />
      </BrowserRouter>,
    );

    expect(getByText(ERROR_MESSAGES.EMPTY_DATA)).toBeInTheDocument();
  });
});
