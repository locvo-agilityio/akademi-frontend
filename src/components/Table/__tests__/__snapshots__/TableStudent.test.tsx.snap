// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableStudent Component Should match snapshot 1`] = `
<div>
  <div
    class="chakra-table__container css-cvn5rz"
  >
    <table
      class="chakra-table css-0"
    >
      <thead
        class="css-0"
      >
        <tr
          class="css-0"
        >
          <th
            class="css-1fjmdth"
          >
            <p
              class="chakra-text css-19uym76"
              title="Name"
            >
              Name
            </p>
          </th>
          <th
            class="css-1fjmdth"
          >
            <p
              class="chakra-text css-19uym76"
              title="ID"
            >
              ID
            </p>
          </th>
          <th
            class="css-1fjmdth"
          >
            <p
              class="chakra-text css-19uym76"
              title="Date"
            >
              Date
            </p>
          </th>
          <th
            class="css-1fjmdth"
          >
            <p
              class="chakra-text css-19uym76"
              title="Parent Name"
            >
              Parent Name
            </p>
          </th>
          <th
            class="css-1fjmdth"
          >
            <p
              class="chakra-text css-19uym76"
              title="City"
            >
              City
            </p>
          </th>
          <th
            class="css-1fjmdth"
          >
            <p
              class="chakra-text css-19uym76"
              title="Contact"
            >
              Contact
            </p>
          </th>
          <th
            class="css-1fjmdth"
          >
            <p
              class="chakra-text css-19uym76"
              title="Grade"
            >
              Grade
            </p>
          </th>
          <th
            class="css-1fjmdth"
          >
            <p
              class="chakra-text css-19uym76"
              title="Action"
            >
              Action
            </p>
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-0"
        >
          <td
            class="css-1g7ry2m"
          >
            <td
              class="css-8faws7"
            >
              <span
                class="chakra-avatar css-1ruhx4w"
              >
                <div
                  aria-label="John Doe"
                  class="chakra-avatar__initials css-0"
                  role="img"
                >
                  JD
                </div>
              </span>
              <p
                class="chakra-text css-tbw3nr"
              >
                John
                 
                Doe
              </p>
            </td>
          </td>
          <td
            class="css-1g7ry2m"
          >
            1
          </td>
          <td
            class="css-1g7ry2m"
          >
            2022-01-01
          </td>
          <td
            class="css-1g7ry2m"
          >
            John Doe
          </td>
          <td
            class="css-1g7ry2m"
          >
            New York
          </td>
          <td
            class="css-1g7ry2m"
          >
            1234567890
          </td>
          <td
            class="css-1g7ry2m"
          >
            A
          </td>
          <td
            class="css-1g7ry2m"
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
